appservice/consumers/roomserver.go

```diff
@@ -123,8 +123,39 @@ func (s *OutputRoomEventConsumer) filterRoomserverEvents(
 	return nil
 }
 
+// appserviceHasMembershipInRoom returns a boolean depending on whether a given
+// appservice has membership at the time a given event was created.
+func (s *OutputRoomEventConsumer) appserviceHasMembershipForEvent(ctx context.Context, event *gomatrixserverlib.HeaderedEvent, appservice config.ApplicationService) bool {
+	// Check if any of the members in the room match the appservice
+	membershipReq := api.QueryStateAfterEventsRequest{
+		PrevEventIDs: []string{event.EventID()},
+		RoomID:       event.RoomID(),
+	}
+	var membershipRes api.QueryStateAfterEventsResponse
+
+	// XXX: This could potentially race if the state for the event is not known yet
+	// e.g. the event came over federation but we do not have the full state persisted.
+	if err := s.rsAPI.QueryStateAfterEvents(ctx, &membershipReq, &membershipRes); err == nil {
```

Comment:
This is the wrong API to use because we're pulling out the entire room state for this call. We can't make use of `QueryStateAfterEventsRequest.StateToFetch` because we want to glob all room member events. Use `QueryMembershipsForRoomRequest` instead:
```go
// QueryMembershipsForRoomRequest is a request to QueryMembershipsForRoom
type QueryMembershipsForRoomRequest struct {
	// If true, only returns the membership events of "join" membership
	JoinedOnly bool `json:"joined_only"`
	// ID of the room to fetch memberships from
	RoomID string `json:"room_id"`
	// ID of the user sending the request
	Sender string `json:"sender"`
}

// QueryMembershipsForRoomResponse is a response to QueryMembershipsForRoom
type QueryMembershipsForRoomResponse struct {
	// The "m.room.member" events (of "join" membership) in the client format
	JoinEvents []gomatrixserverlib.ClientEvent `json:"join_events"`
	// True if the user has been in room before and has either stayed in it or
	// left it.
	HasBeenInRoom bool `json:"has_been_in_room"`
	// True if the user asked to forget this room.
	IsRoomForgotten bool `json:"is_room_forgotten"`
}
```

Reply:
`QueryMembershipsForRoomRequest` doesn't work here because it takes a `sender` param, and we don't have a sender here.